### DISCUSSION TITLE
TTGO T7 and PCM5102: working

### DISCUSSION THREAD

**Discussion by Joshfindit** [sle118/squeezelite-esp32#186] [Show and tell]
*This does go in to detail about how I got this working as someone who had zero experience with squeezelite. Hopefully it helps others!*

# Prerequisites

- ESP32
    As mentioned in the [README](https://github.com/sle118/squeezelite-esp32/blob/master-cmake/README.md); you can't use the cheapest ESP32 boards/module. You need at least 4MB of PSRAM. Easy option is to save WROOM modules for other projects and grab a WROVER.
- PCM5102 I2S DAC
    This is a decent and cheap chip that is easily available on PCBs
- USB power
- Something to plug in to the headphone jack. This can be earbuds/other low-power headphones, or an amp that can power some speakers (and the speakers themselves).

# Step 1: Prep the PCM5102 PCB

I used (and I suspect many others might use) something like this:
![Sac1b92d17e5f4baa8a78673e752aef865](https://user-images.githubusercontent.com/1937155/193966175-62565219-135c-443c-a692-735bb0ee2257.jpg)

![S0212dca774134de4a9da1f16972ab741d](https://user-images.githubusercontent.com/1937155/193966053-ad716c7c-3dab-4e22-8076-61c8595cf015.jpg)

But there is a possibility that yours will not work right away.

[This helpful user](https://github.com/dwhinham/mt32-pi/discussions/46) led me to uncover why _my_ PCB wasn't working: the solder pads were not connected properly.

Either test the connections with a multimeter (The blob should read 3.3V for "H3L" and be directly connected to ground for "H1L", "H2L", and "H4L".

I redid all the blobs fresh, and that worked.

# Step 2: Wire up the wires

The README shows

| [Pin on] I2S | [Pin on] WROVER |
|---|---|
|VCC | 3.3V|
|3.3V | 3.3V|
|GND | GND|
|FLT | GND|
|DMP | GND|
|SCL | GND|
|BCK | (BCK - see below)|
|DIN | (DO - see below)|
|LCK | (WS - see below)|
|FMT | GND|
|XMT | 3.3V|

However, at least for this PCB; this is confusing.

## How mine got wired

| Pin on I2S PCB | Pin on TTGO T7 |
|---|---|
| VIN | 3.3V|
|GND | GND|
|FLT | N/A: FLT is Pin 1, and when "H1L" is bridged between `1` and `L`, it's connected to ground|
|DMP | N/A: This is labelled `DEMP`, is pin 2, and "H2L" is bridged between `2` and `L`|
|SCL | N/A: No connection is labelled SCL or anything like that |
|FMT | N/A: FMT is Pin 4, and "H4L" is bridged between `4` and `L`|
|XMT | N/A: Nothing labelled XMT. There is XSMT which mutes the chip entirely unless it's connected to 3.3V, but this is "H3L" and `H` and `3` are bridged |
| BCK | GPIO21 |
| DIN | GPIO22 |
| LCK (Actually labelled LRCK) | GPIO27 |

## Notes

- It doesn't seem to matter which GPIO pins you use for I2S, and I'm going to change them shortly so that they are all connected in a row on the same side as 3.3V, but these worked for me.
- Someone mentioned that you can connect the I2S PCB's VIN to 3-6V, so I connected it to 5V thinking more power=better, BUT it clearly cannot handle 5V. It was crunchy sounding through the headphones. Use 3.3V.
- "See below"
  - There's apparently some industry confusion about what to call the various I2S pins, and so that's how you get things like "LCK is actually LRCK.. And WS"
  - As with FTDI, DIN (Data In) on one device connects to DO (Data Out) on the other.

# The config

## What to know

Once your board is flashed, powered, connected to your wifi, and you are viewing the config page in your browser:

- You can find the NVS Editor by going to "Credits" then checking "Show NVS Editor" at the bottom of the page
- NVS Editor is apparently the more reliable and advanced editor. Even if you're new you'll probably want to start here as most people who help you will direct you to use it during troubleshooting.

### Option 1: use the "Hardware" tab under "DAC Options"

1. Change "DAC model Name" to `I2S`
2. Set "Clock GPIO" to the ESP32 GPIO pin connected to `BCK` on the I2S PCB
3. Set "Word Select GPIO" to the ESP32 GPIO pin connected to `LRCK` on the I2S PCB
4. Set "Data GPIO" to the ESP32 GPIO pin connected to `DIN` on the I2S PCB
5. Save and Apply (Actually, in my case "Save" didn't work as well as "Apply", but I don't know the repercussions for using one over the other)

### Option 2: Use the NVS Editor

1. Scroll down to the `dac_config` key
3. Enter the value of `model=I2S,bck=<GPIO for BCK>,ws=<GPIO for LRCK>,do=<GPIO for DIN>`
     Such as `model=I2S,bck=21,ws=27,do=22`
4. Press "Commit"

# Success! (hopefully)

Reboot if needed, connect to it with your phone/tablet/laptop/whatever and start playing music.
You should hear the sound come through clearly and without being sped up/slowed down.

Play your favourite track and celebrate 🎉

Or if you're like me: go to bed because you've spent way too long making this work and you have stuff to do in the morning.

**Comment by tim-peters** [comment]
Thank you very much! 
Is there any way other than actually hearing something to find out if the i2s connection is working? For example, does the esp output something like "i2s connection successfully established" on the serial output when booting?

**Reply by sle118** [reply]
I2S is not bi directional so no. The gpio's are pretty much "shouting" signal regardless if there is anything listening. 

@Joshfindit thanks for the nice write-up

**Comment by nerdydig** [comment]
Thanks so much for making this. I was running around in circles trying to figure out how to get these two boards working together. Did you happen to set things up to get this working with Spotify? I have been trying to get this to show in LMS without any luck.

**Comment by orgebar** [comment]
Nice, thanks!

**Comment by lisz8** [comment]
Thank you for this! I've got a similar ESP32 and similar DAC (from Adafruit). Then I looked at that README page and got super confused. Works like a charm! :)

**Comment by cmonrroy88** [comment]
Thank you so much. It's just the guide that I was searching for many days.

**Comment by p1ne** [comment]
In my case wiring SCK from PCM5102 to GND was mandatory to get proper sound
